File: app/src/main/java/org/qp/android/questopiabundle/utils/FileUtil.java
package org.qp.android.questopiabundle.utils;

import android.content.Context;
import android.net.Uri;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.documentfile.provider.DocumentFile;

import com.anggrayudi.storage.FileWrapper;
import com.anggrayudi.storage.file.CreateMode;
import com.anggrayudi.storage.file.DocumentFileUtils;

import java.io.ByteArrayOutputStream;
import java.io.IOException;

public final class FileUtil {

    public static boolean isWritableFile(Context context, DocumentFile file) {
        if (file == null) return false;
        var canWrite = DocumentFileUtils.isWritable(file, context);
        return file.exists() && file.isFile() && canWrite;
    }

    @Nullable
    public static byte[] getFileContents(@NonNull Context context,
                                         @NonNull Uri uriContent) {
        var resolver = context.getContentResolver();
        try (var in = resolver.openInputStream(uriContent);
             var out = new ByteArrayOutputStream()) {
            if (in != null) {
                StreamUtil.copy(in, out);
            } else {
                throw new NullPointerException();
            }
            return out.toByteArray();
        } catch (Exception ex) {
            // TODO: 04.12.2024 Add logger
            Log.e("FileUtil", "Error reading file: " + uriContent, ex);
            return null;
        }
    }

    public static FileWrapper.Document documentWrap(DocumentFile inputFile) {
        return new FileWrapper.Document(inputFile);
    }

    public static void writeFileContents(@NonNull Context context,
                                         @NonNull Uri uriContent,
                                         byte[] dataToWrite) {
        var resolver = context.getContentResolver();
        try (var out = resolver.openOutputStream(uriContent, "w")) {
            if (out != null) {
                out.write(dataToWrite);
            } else {
                throw new IOException("Input is NULL!");
            }
        } catch (IOException ex) {
            // TODO: 04.12.2024 Add logger
            Log.e("FileUtil", "Error reading file: " + uriContent, ex);
        }
    }

    @Nullable
    public static DocumentFile findOrCreateFile(Context context,
                                                DocumentFile srcDir,
                                                String name,
                                                String mimeType) {
        return DocumentFileUtils.makeFile(srcDir, context, name, mimeType, CreateMode.REUSE);
    }

    public static DocumentFile fromRelPath(@NonNull Context context,
                                           @NonNull final String path,
                                           @NonNull DocumentFile parentDir,
                                           final boolean requiresWriteAccess) {
        return DocumentFileUtils.child(parentDir, context, path, requiresWriteAccess);
    }

    @Nullable
    public static DocumentFile fromFullPath(@NonNull Context context,
                                            @NonNull String fullPath,
                                            @NonNull DocumentFile rootDir) {
        var findDir = rootDir;
        var nameGameDir = rootDir.getName();

        try {
            var index = fullPath.indexOf(nameGameDir);
            var subString = fullPath.substring(index);
            var splitString = subString.replace(nameGameDir + "/", "");
            var pathToFileSegments = splitString.split("/");

            for (var segment : pathToFileSegments) {
                if (segment.isEmpty()) {
                    continue;
                }
                findDir = fromRelPath(context, segment, findDir, true);
                if (findDir == null) {
                    break;
                }
            }
        } catch (NullPointerException i) {
            return null;
        }

        return findDir;
    }
}

File: app/src/main/java/org/qp/android/questopiabundle/utils/FileUtil.kt
package org.qp.android.questopiabundle.utils

import android.content.Context
import android.net.Uri
import android.util.Log
import androidx.documentfile.provider.DocumentFile
import com.anggrayudi.storage.FileWrapper
import com.anggrayudi.storage.file.DocumentFileCompat
import com.anggrayudi.storage.file.child
import com.anggrayudi.storage.file.isWritable
import org.qp.android.questopiabundle.utils.StreamUtil.copy
import java.io.ByteArrayOutputStream
import java.io.IOException
import kotlin.contracts.ExperimentalContracts
import kotlin.contracts.contract

object FileUtil {
    @ExperimentalContracts
    fun isWritableFile(context: Context, file: DocumentFile?): Boolean {
        contract {
            returns() implies (file != null)
        }
        if (file == null) return false
        return file.exists() && file.isFile && file.isWritable(context)
    }

    @ExperimentalContracts
    fun isWritableDir(context: Context, dir: DocumentFile?): Boolean {
        contract {
            returns() implies (dir != null)
        }
        if (dir == null) return false
        return dir.exists() && dir.isDirectory && dir.isWritable(context)
    }

    fun getFileContents(
        context: Context,
        uriContent: Uri
    ): ByteArray? {
        val resolver = context.contentResolver
        try {
            resolver.openInputStream(uriContent).use { `in` ->
                ByteArrayOutputStream().use { out ->
                    if (`in` != null) {
                        copy(`in`, out)
                    } else {
                        throw NullPointerException()
                    }
                    return out.toByteArray()
                }
            }
        } catch (ex: Exception) {
            // TODO: 04.12.2024 Add logger
            Log.e("FileUtil", "Error reading file: $uriContent", ex)
            return null
        }
    }

    fun documentWrap(inputFile: DocumentFile): FileWrapper.Document {
        return FileWrapper.Document(inputFile)
    }

    fun writeFileContents(
        context: Context,
        uriContent: Uri,
        dataToWrite: ByteArray?
    ) {
        val resolver = context.contentResolver
        try {
            resolver.openOutputStream(uriContent, "w").use { out ->
                if (out != null) {
                    out.write(dataToWrite)
                } else {
                    throw IOException("Input is NULL!")
                }
            }
        } catch (ex: IOException) {
            // TODO: 04.12.2024 Add logger
            Log.e("FileUtil", "Error reading file: $uriContent", ex)
        }
    }
}
